shopping-cart/shopping-cart-scala/shopping-cart/src/main/scala/com/example/shoppingcart/impl/ShoppingCart.scala

```diff
@@ -0,0 +1,250 @@
+      if ((json \ "reason").isDefined)
+        Json.fromJson[Rejected](json)
+      else
+        Json.fromJson[Accepted](json)
+    }
+
+    override def writes(o: Confirmation): JsValue = {
+      o match {
+        case acc: Accepted => Json.toJson(acc)
+        case rej: Rejected => Json.toJson(rej)
+      }
+    }
+  }
+
+  val empty = ShoppingCart(Map.empty, checkedOut = false)
+
+  val typeKey: EntityTypeKey[Command] = EntityTypeKey[Command]("ShoppingCart")
+
+  // We can then access the entity behavior in our test tests, without the need to tag
+  // or retain events.
+  def behavior(persistenceId: PersistenceId): EventSourcedBehavior[Command, Event, ShoppingCart] = {
+    EventSourcedBehavior
+      .withEnforcedReplies[Command, Event, ShoppingCart](
+        persistenceId = persistenceId,
+        emptyState = ShoppingCart.empty,
+        commandHandler = (cart, cmd) => cart.applyCommand(cmd),
+        eventHandler = (cart, evt) => cart.applyEvent(evt)
+      )
+  }
+
+  def behavior(entityContext: EntityContext[Command]): Behavior[Command] =
+    behavior(PersistenceId(entityContext.entityTypeKey.name, entityContext.entityId))
+      .withTagger(AkkaTaggerAdapter.fromLagom(entityContext, Event.Tag))
+      .withRetention(RetentionCriteria.snapshotEvery(numberOfEvents = 100, keepNSnapshots = 2))
+
+  /**
+   * The aggregate get snapshoted every configured number of events. This
+   * means the state gets stored to the database, so that when the entity gets
+   * loaded, you don't need to replay all the events, just the ones since the
+   * snapshot. Hence, a JSON format needs to be declared so that it can be
+   * serialized and deserialized when storing to and from the database.
+   */
+  implicit val format: Format[ShoppingCart] = Json.format
+}
+
+final case class ShoppingCart(items: Map[String, Int], checkedOut: Boolean) {
+
+  import ShoppingCart._
+
+  //The shopping cart behavior changes if it's checked out or not. The command handles are different for each case.
+  def applyCommand(cmd: Command): ReplyEffect[Event, ShoppingCart] =
+    if (checkedOut) {
+      cmd match {
+        // it is allowed to read it's state of a checked-out cart
+        case Get(replyTo) => Effect.reply(replyTo)(toSummary(this))
+        // CheckedOut is a final state, no mutations allowed
+        case AddItem(_, _, replyTo) => Effect.reply(replyTo)(Rejected("Cannot add an item to a checked-out cart"))
+        case RemoveItem(_, replyTo) => Effect.reply(replyTo)(Rejected("Cannot remove an item from a checked-out cart"))
+        case AdjustItemQuantity(_, _, replyTo) =>
+          Effect.reply(replyTo)(Rejected("Cannot adjust an item quantity on a checked-out cart"))
+        case Checkout(replyTo) => Effect.reply(replyTo)(Rejected("Cannot checkout a checked-out cart"))
+      }
+    } else {
+      cmd match {
+        case AddItem(itemId, quantity, replyTo)            => addItem(itemId, quantity, replyTo)
```

Comment:
This style is used in Akka docs and if I remember correctly @renatocaval was advocating it as well. I moved everything to `applyCommand` to avoid adding more hops though.